### ISSUE TITLE
Backport Altering - cluster-altering app shows error after upgrading Rancher

### ISSUE THREAD

**Issue by alena1108** [rancher/rancher#21364]
Backport https://github.com/rancher/rancher/issues/21323

**Comment by jiaqiluo** [comment]
Failed to validate the bug fix.

Steps:
- run Rancher: v2.2.4 container
- as admin, enable Github auth
- log in as Github user user1
- add the following three clusters: DO single-node, DO 3-node, and GKE 
- adds cluster notifier and cluster alerting for all three clusters
- upgrade Rancher to v2.2.5-rc9

Results:
- in the clusters DO single-node and GKE, the app `cluster-alerting` and `monitoring-operator` get upgraded successfully, meaning there is only 1 workload left under the app `cluster-alerting`, and the altering feature is functional 
- in the cluster DO 3-node, the app `cluster-alerting` does NOT get upgraded, meaning there are still 2 workloads under the app  `cluster-alerting`, but the cluster altering still works
- and Rancher's log shows  
```
2019/07/08 20:56:00 [ERROR] ProjectController c-44pq6/p-jmpkk [system-image-upgrade-controller] failed with : upgrade system service alerting failed, cluster c-44pq6 not ready
2019/07/08 20:56:00 [ERROR] ProjectController c-44pq6/p-scrx7 [system-image-upgrade-controller] failed with : upgrade system service alerting failed, cluster c-44pq6 not ready
```


More Info 1
- adding a new single-node cluster and a new 3-node cluster in the same setup and enabling the cluster altering fails
- Rancher shows the following errors
```
Failed to ensure monitoring project name: failed to move Namespace cattle-prometheus into a Project: Operation cannot be fulfilled on namespaces "cattle-prometheus": the object has been modified; please apply your changes to the latest version and try again
```
- Rancher's log shows
 
```
2019/07/08 21:31:58 [ERROR] ClusterAlertGroupController c-28fpf/cag-8cwmg [cluster-alert-group-deployer] failed with : failed to get Alertmanager Deployment information, statefulsets.apps "alertmanager-cluster-alerting" not found, [cluster-alert-group-controller] failed with : Failed to get service for alertmanager, service "cattle-prometheus/alertmanager-operated" not found
2019/07/08 21:32:00 [ERROR] ClusterAlertRuleController c-28fpf/car-8fw8v [cluster-alert-rule-deployer] failed with : failed to get Alertmanager Deployment information, statefulsets.apps "alertmanager-cluster-alerting" not found, [cluster-alert-rule-controller] failed with : Failed to get service for alertmanager, service "cattle-prometheus/alertmanager-operated" not found
2019/07/08 21:32:04 [ERROR] ClusterAlertGroupController c-28fpf/cag-8cwmg [cluster-alert-group-deployer] failed with : failed to get Alertmanager Deployment information, statefulsets.apps "alertmanager-cluster-alerting" not found, [cluster-alert-group-controller] failed with : Failed to get service for alertmanager, service "cattle-prometheus/alertmanager-operated" not found
```


<img width="1408" alt="screenshot" src="https://user-images.githubusercontent.com/6218999/60845866-7ca50f00-a192-11e9-8dd4-09d8c849480a.png">

More Info 2
- tried the same steps in a fresh-installed Rancher v2.2.5-rc9 setup and the cluster altering works with no issues
- tried the same steps in a fresh-installed Rancher v2.2.5-rc8 setup and the cluster altering works with no issues

**Comment by aiwantaozi** [comment]
Hi, the original issue https://github.com/rancher/rancher/issues/21323 
***Problem***: Upgraded the rancher server, even the cluster is not ready but the cluster-alerting app continue to upgrade,  then get error ```waiting for cluster agent to connect```

***Solution***: Check cluster status, only upgrade when the cluster status is ready, if not ready will throw error `upgrade system service alerting failed, cluster c-44pq6 not ready`, then rancher will retry. So in the first `DO 3-node` you see the log `[system-image-upgrade-controller] failed with : upgrade system service alerting failed, cluster c-44pq6 not ready`. This error should not happen after the cluster is ready and retry upgrade, the upgrade should be ok. ***Whether the cluster-alerting work after a while(auto error retry happened)?***

***For the Update 1 above***: Because we don't release new charts to fix this two operators issue, but fixed by update the app's, so it still has 2 workloads if you directly upgrade the app `cluster-alerting` in the UI, it's expected now.

Others error may not be related to this PR, but I will also try to reproduce, thank you

**Comment by sangeethah** [comment]
@aiwantaozi   , once we see this error in the logs when cluster alerting upgrade was attempted and it fails , `[system-image-upgrade-controller] failed with : upgrade system service alerting failed, cluster c-44pq6 not ready` , do you expect the upgrade to be attempted again and succeed at some point ?

This never happened . The  the app cluster-alerting does NOT get upgraded, meaning there are still 2 workloads under the app cluster-alerting. 

`For the Update 1 above: Because we don't release new charts to fix this two operators issue, but fixed by update the app's, so it still has 2 workloads if you directly upgrade the app cluster-alerting in the UI, it's expected now.` ==> There was no attempt made to update "cluster-alerting"  the app from UI .

"adding a new single-node cluster and a new 3-node cluster in the same setup and enabling the cluster altering fails" ==> Is there any chance that the earlier attempt to auto upgrade cluster alert being prevented( because of cluster state not being "active")  result in enabling alerts in other cluster having a problem ?

**Comment by aiwantaozi** [comment]
Find a history problem will cause some clusters upgrade system services failed when there are multiple clusters, it's included since v2.1.0.

Root cause:
this cause by only one alert/logging/pipeline upgradeService was inited, when there are multiple clusters, the former clusterName will be overwritten by the laster one, so only one cluster's system service is upgraded success. 

Steps to reproduce:
run Rancher: v2.2.4 container
add the following three clusters: DO single-node, DO 3-node
adds cluster notifier and cluster alerting for all three clusters
upgrade Rancher to v2.2.5-rc9
only one cluster upgrade success

**Comment by loganhz** [comment]
Please create another issue to check `More Info 1` section. It’s a separate issue.

**Comment by alena1108** [comment]
The decision was made to not put this fix to 2.2.5 due to a chance of regressions. Will be fixed in 2.3

**Comment by jiaqiluo** [comment]
> Please create another issue to track issues in `More Info 1` section. It’s a separate issue.

it is tracked here: https://github.com/rancher/rancher/issues/21399

**Comment by jiaqiluo** [comment]
From @loganhz 

If you hit this issue in 2.2.4 or 2.2.5, please upgrade the cluster-alerting app and add operator.enabled=false answer manually to workaround it.

It will get fixed in 2.3.

FYI, when you hit this issue only the state of the cluster app is in red. But alerting feature is still functional in your cluster.
